src/store/uploads/uploads.provider.js

```diff
@@ -0,0 +1,64 @@
+import { storage } from '@/firebase'
+import { ref, uploadBytesResumable } from 'firebase/storage'
+
+const variants = {
+  sucess: 'success',
+  error: 'error',
+  loading: 'loading',
+}
+
+export const startUpload = (setUploads, files) => {
+  files.forEach((file, currentIndex) => {
+    // TODO: show error message if file type is not allowed
+
+    const songsRef = ref(storage, `songs/${file.name}`)
+    const task = uploadBytesResumable(songsRef, file)
+
+    const newUpload = {
+      task,
+      currentProgress: 0,
+      name: file.name,
+      variant: variants.loading,
+      icon: 'fas fa-spinner fa-spin',
+    }
+
+    setUploads((prev) => [newUpload, ...prev])
+
+    task.on(
+      'state_changed',
+      (snapshot) => {
+        console.log(snapshot)
+        const progress = (snapshot.bytesTransferred / snapshot.totalBytes) * 100
+
+        // Update progress in the state for the current upload
+        setUploads((prevUploads) => {
```

Comment:
quick solution: you can move the uploads local state to the slice state, and you can create different actions to trigger, try to split each step of this function, move the logic to the thunk, which is like the 'manager', and for the providers only create functions that interact with firebase.

Reply:
For example, the uploadResumableBytes method from firebase could be a function that takes a file and return a task. when with that result you can create another function for handling the uploading progress.

variant, icon, progress, etc, because those fields does not belong to firebase, see if you can move this logic outside of the provider.